src/react/tests/useQuery.test.tsx

```diff
@@ -817,10 +890,11 @@ describe('useQuery', () => {
 
   it('should re-render when dataUpdatedAt changes but data remains the same', async () => {
     const key = queryKey()
-    const states: UseQueryResult<string>[] = []
+    const states: UseQueryResult<number>[] = []
+    let queryReturnValue = 0
 
     function Page() {
-      const state = useQuery(key, () => 'test', {
+      const state = useQuery(key, () => queryReturnValue++, {
```

Comment:
this test is named:

`'should re-render when dataUpdatedAt changes but data remains the same'`

now, data no longer stays the same. I was wondering if we're not breaking this functionality: Apparently, we would still expect `dataUpdatedAt` to be updated if a new fetch comes in, which now likely doesn't happen because of the bail out...

---

sorry for the troubles, but I think the easiest solution would be to remove the `Object.is` related bailouts and stick with the `undefined` ones?

Reply:
Your instinct to combine undefined and same-data bailout was good, since the code changes happen in the same places. Let's stick with this. It is not hard to add code to update dataUpdatedAt. I just need to enhance the 'ignore' action handler to do this.

Reply:
Hm, I don't really think so. When I wrote [this suggestion](https://github.com/tannerlinsley/react-query/discussions/2823#discussioncomment-1562184), I only had `setQueryData` in mind, because that's where the issue stems from: That you cannot bail out of it even if you get `undefined` passed in and can't create valid data.

I really think we should keep all changes confined to `setQueryData`. If we fetch data, we can just put it in the cache without bailing out. structural sharing will make sure that data stays referentially stable, and RQ won't even re-render observers if the data doesn't change. So what's the point in bailing out here? I think it would also remove lots of code.
Additionally, we _should_ in fact call the `onSuccess` callback of observers that the fetch was successful, even if the data stays the same. I think bailing out here does more harm than good.

Reply:
The structural sharing takes place in `Query.setData`, which is used by `setQueryData` and fetching. So if we are going to give up on the Object.is bailout for fetching, we should do so for setQueryData as well.

Similarly, the call to onSuccess takes place in `Query.setData` as well. So if we want to return onSuccess for fetching but not setQueryData, it would require some kind of flag, which is smelly.

Since these changes are for React Query 4, I am going to venture my opinion. A large part of the confusion here is that we are using `onSuccess` in two very different ways. We use it to indicate a successful fetch, and we use it to indicate a data change. Instead, we need two different callbacks, `onSuccess` when a fetch succeeds and `onDataChanged` whenever cache is updated with new data via `setQueryData` or the fetch. I can move this discussion to Discussions if that seems more appropriate.

Reply:
> The structural sharing takes place in Query.setData, which is used by setQueryData and fetching.

this is what I meant when I said we'd need some internal restructurings. calling the functional updater + performing structural sharing / isDataEqual check can be a pure function - it only depends on `data` and `options` - both of which we have available. This only leaves us with

```
    // Set data and mark it as cached
    this.dispatch({
      data,
      type: 'success',
      dataUpdatedAt: options?.updatedAt,
    })
```

that we would need to somehow put into a function on the query, like `setSuccessData` and call it in two places: from `query.setData` and from `queryClient.setQueryData`, where we could only do it after we've computed the new data by calling the aforementioned function, so potentially, not calling `query.setData` from `queryClient.setQueryData`, thus keeping the separation between "setting data after fetching" and "setting data from setQueryData"

---

i like the idea of an extra callback - it was confusing to ppl that `onSuccess` is also called after `setQueryData`, but it's a different topic I think

Reply:
Ok, this all makes sense. I will get to work on it.

If we ever split out onDataChanged and onSuccess, we will need to do the kind of code split you are proposing above anyway. I will propose this change that in a discussion.

Only one issue to resolve: what do we do if the query function returns undefined? If we are not bailing out, I suggest we treat this as an error.

Reply:
Proposed onDataChanged callback: https://github.com/tannerlinsley/react-query/discussions/2911

Reply:
> Only one issue to resolve: what do we do if the query function returns undefined? If we are not bailing out, I suggest we treat this as an error.

this is also what Tanner proposed yesterday for v4: we throw an error and put the query in error state. In development mode, we could additionally warn, and if possible, we should exclude undefined on type level :)

Reply:
Sounds good, I will make this change.